message/deduplication/algorithm.go

```diff
@@ -0,0 +1,102 @@
+package deduplication
+
+import "fmt"
+
+type bloomFilterBackend interface {
+	Set(key string) error
+	Exists(key string) (bool, error)
+	CheckAndSet(key string) (bool, error)
+}
+
+type bloomFilterOptions struct {
+	validators []KeyValidator
+}
+
+type bloomFilterOptionsFunc struct {
+	f func(o *bloomFilterOptions)
+}
+
+func (f bloomFilterOptionsFunc) apply(o *bloomFilterOptions) {
+	f.f(o)
+}
+
+type BloomFilterOption interface {
+	apply(*bloomFilterOptions)
+}
+
+func WithBloomFilterKeyValidators(vs ...KeyValidator) BloomFilterOption {
+	return bloomFilterOptionsFunc{
+		f: func(o *bloomFilterOptions) {
+			o.validators = vs
```

Comment:
Maybe should be `append(o.validators, vs...)`? I think that would be more consistent with other option types in leg.